llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp

```diff
@@ -902,6 +904,109 @@ struct GeneralizePackOpPattern : OpRewritePattern<PackOp> {
   }
 };
 
+/// Rewrites iree_linalg_ext.unpack to rank-reduced extract_slice op + transpose
+/// op + insert_slice op. It requires the outer dims are all 1s.
+struct GeneralizeUnPackOpPattern : OpRewritePattern<UnPackOp> {
+  using OpRewritePattern<UnPackOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(UnPackOp unpackOp,
+                                PatternRewriter &rewriter) const final {
+    if (!unpackOp.hasTensorSemantics()) {
+      return rewriter.notifyMatchFailure(unpackOp, "require tensor semantics");
+    }
+
+    int64_t outputRank = unpackOp.getOutputRank();
+    if (llvm::any_of(unpackOp.getInputShape().take_front(outputRank),
+                     [](int64_t val) { return val != 1; })) {
+      return rewriter.notifyMatchFailure(
+          unpackOp, "require the outer dimension of the result are all 1s");
+    }
+
+    int64_t inputRank = unpackOp.getInputRank();
+
+    Location loc = unpackOp.getLoc();
+    Attribute zeroIdxAttr = rewriter.getIndexAttr(0);
+    Attribute oneIdxAttr = rewriter.getIndexAttr(1);
+    SmallVector<OpFoldResult> readOffsets(inputRank, zeroIdxAttr);
+    SmallVector<OpFoldResult> readStrides(inputRank, oneIdxAttr);
+
+    auto mixedTiles = unpackOp.getMixedTiles();
+    SmallVector<OpFoldResult> readSizes(outputRank, oneIdxAttr);
+    readSizes.append(mixedTiles.begin(), mixedTiles.end());
+
+    // Explicitly create the type because the inner tile size could be 1. We
+    // want to represent the whole inner tile in this case.
+    SmallVector<int64_t> readShape;
+    for (auto sz : mixedTiles) {
+      auto cst = getConstantIntValue(sz);
+      readShape.push_back(cst ? cst.value() : ShapedType::kDynamicSize);
+    }
+    Type elemType = unpackOp.getInputType().getElementType();
+    auto readType = RankedTensorType::get(readShape, elemType);
+    Value innerTile = rewriter.create<tensor::ExtractSliceOp>(
+        loc, readType, unpackOp.getInput(), readOffsets, readSizes,
+        readStrides);
+
+    SmallVector<int64_t> innerDimsPos =
+        extractFromI64ArrayAttr(unpackOp.getInnerDimsPos());
+    auto interchangeVector =
+        computeInterchangeFromDimPos(innerDimsPos, outputRank);
+    SmallVector<int64_t> transpShape =
+        interchange<int64_t>(readShape, interchangeVector);
+    auto transpType = RankedTensorType::get(transpShape, elemType);
```

Comment:
I think you should be able to use `PackOp::getPackedType` here...

Reply:
The getPackedType will return the type of input type. What we want here is the result type of transpose(inner tile).

The rank of output is less than the rank of input in unpack op. While not every outer dim maps to an inner dim, we have to explicitly compute the transpType of inner dim, and insert them to the corresponding position.

Take the first test case as example:

```
// simple_KCRSsr_to_KCRS
%arg0: tensor<1x1x1x1x8x32xf32>, %arg1: tensor<1x1x32x8xf32>
```

First of all, we extract the inner tile `8x32`.

We have to transpose it (for mapping sr to RS) to `32x8`. The `32x8` can only be computed by interchanging the readShape (`8x32`).

=====

After studying the code a bit more, I found that I can use `getInputShape().drop_front(outputRank)` for readShape to simplify the code a bit.